src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,478 @@
+				'META' === $tag_name ||
+				'NOFRAMES' === $tag_name ||
+				'SCRIPT' === $tag_name ||
+				'STYLE' === $tag_name ||
+				'TEMPLATE' === $tag_name ||
+				'TITLE' === $tag_name
+			) ||
+			(
+				'closer' === $tag_type &&
+				'TEMPLATE' === $tag_name
+			) )
+		{
+			parent::seek( 'current' );
+			$this->insertion_mode = 'in-head';
+			return $this->step();
+		}
+
+		/*
+		 * > A start tag whose tag name is "body"
+		 */
+		if ( 'opener' === $tag_type && 'BODY' === $tag_name ) {
+			goto ignored;
+		}
+
+		/*
+		 * > A start tag whose tag name is "frameset"
+		 */
+		if ( 'opener' === $tag_type && 'FRAMESET' === $tag_name ) {
+			return self::NOT_IMPLEMENTED_YET;
+		}
+
+		/*
+		 * > An end-of-file token
+		 *
+		 * Stop parsing.
+		 */
+
+		/*
+		 * > An end tag whose tag name is "body"
+		 * > An end tag whose tag name is "html"
+		 */
+		if ( 'closer' === $tag_type && ( 'BODY' === $tag_name || 'HTML' === $tag_name ) ) {
+			/*
+			 * > If the stack of open elements does not have a body element in scope, this is a parse error; ignore the token.
+			 *
+			 * @TODO: We didn't construct an open HTML or BODY tag, but we have to make a choice here based on that.
+			 *        Probably need to create these _or_ assume this will always transfer to "after body".
+			 */
+			$this->insertion_mode = 'after-body';
+			return true;
+		}
+
+		/*
+		 * > A start tag whose tag name is one of: "address", "article", "aside", "blockquote", "center",
+		 * > "details", "dialog", "dir", "div", "dl", "fieldset", "figcaption", "figure", "footer",
+		 * > "header", "hgroup", "main", "menu", "nav", "ol", "p", "search", "section", "summary", "ul"
+		 */
+		if (
+			'opener' === $tag_type && (
+				'ADDRESS' === $tag_name ||
+				'ARTICLE' === $tag_name ||
+				'ASIDE' === $tag_name ||
+				'BLOCKQUOTE' === $tag_name ||
+				'CENTER' === $tag_name ||
+				'DETAILS' === $tag_name ||
+				'DIALOG' === $tag_name ||
+				'DIR' === $tag_name ||
+				'DIV' === $tag_name ||
+				'DL' === $tag_name ||
+				'FIELDSET' === $tag_name ||
+				'FIGCAPTION' === $tag_name ||
+				'FIGURE' === $tag_name ||
+				'FOOTER' === $tag_name ||
+				'HEADER' === $tag_name ||
+				'HGROUP' === $tag_name ||
+				'MAIN' === $tag_name ||
+				'MENU' === $tag_name ||
+				'NAV' === $tag_name ||
+				'OL' === $tag_name ||
+				'P' === $tag_name ||
+				'SEARCH' === $tag_name ||
+				'SECTION' === $tag_name ||
+				'SUMMARY' === $tag_name ||
+				'UL' === $tag_name
+			)
+		) {
+			if ( $this->has_in_scope( 'P', 'BUTTON' ) ) {
+				$this->close_p_element();
+			}
+
+			$this->enter_element( $tag_name );
+		}
+
+		return false;
+	}
+
+	public function set_bookmark( $bookmark_name ) {
+		return parent::set_bookmark( '_' . $bookmark_name );
+	}
+
+	public function release_bookmark( $bookmark_name ) {
+		return parent::release_bookmark( '_' . $bookmark_name );
+	}
+
+	private function enter_element( $element ) {
+		$this->depth++;
+
+		parent::set_bookmark( "{$this->depth}_{$element}" );
```

Comment:
also, why set a bookmark for each element?

Reply:
this is a stub function for the stack of open elements. I think I'm going to rip it out and use a proper stack, but we have to have a bookmark _because_ if we make edits we have to know that the stack might have changed.

Reply:
the stack of open elements might actually shift and some may disappear. in any case, if we `seek()` we have to reset them, so we need a good way to reconstruct the stack when seeking.

in your PR I think you're doing that by storied copies of all the open elements for every bookmark, but also we have to make sure those are bookmarks _because_ we expect changes to be applied to the document which could change things.

the main case that matters I think is wiping out a bookmark, but inserting nodes will shift them all.

```html
<div><p>inside</p></div>
     ^^^ bookmark

unwrap

<p>inside</p>
^^^ stack of open elements here is no longer DIV > P but rather only P
```